### PR TITLE
[WIP-FEAT] MoM Event (REBASED)

### PR DESCRIPTION
**Pull request by ShiningMinions**
# Description

## IMPORTANT NOTES:

- This is the REBASED version of original PR https://github.com/sunflower-land/sunflower-land/pull/982. That PR has been closed in favor of this one.
- @JacobSobolev 's https://github.com/sunflower-land/sunflower-land/pull/986 is a subset of this PR. Perhaps we should close his PR, now that this PR is much cleaner than the original.
- I don't rebase often, so PLEASE double-check history to make sure everything is looking as expected.

This PR contains the community contributions for the MoM event. This is a WIP, I will update this description as needed.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

WIP - Needs testing.

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [ ] Screenshot if it includes any UI changes
- [x] I have read the contributing guidelines and agree to the T&Cs
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by ShiningMinions** [comment]
@adamhannigan @jcads @JacobSobolev If accepted, we will need to start working in my new rebased branch here: https://github.com/ShiningMinions/sunflower-land/tree/feat/mom

I have invited you all to my forked repo, so you should have direct access to this branch now.

File: src/features/goblins/rocket/Rocket.tsx [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+      );
+    }
+
+    if (hasCompletedMission) {
+      return (
+        <>
+          <span className="text-shadow block my-2 text-xs sm:text-sm">
+            Great job on Mars Interplanetary Farmer! In exchange for your token
+            received on Mars, I will give you something to remember me by. After
+            this trade, go back to your farm and sync on chain to see it.t.
```
Comment (brian-man) [inline src/features/goblins/rocket/Rocket.tsx:155]: `it.t.` -> `it.`

File: src/features/goblins/rocket/Rocket.tsx [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+      context: { state },
+    },
+  ] = useActor(goblinService);
+  const [isDialogOpen, setIsDialogOpen] = useState(false);
+  const [isItemsOpen, setIsItemsOpen] = useState(false);
+  const [isRocketLaunching, setIsRocketLaunching] = useState(false);
+  const [isRocketLaunchComplete, setIsRocketLaunchComplete] = useState(false);
+  const [hasCompletedMission, setHasCompletedMission] = useState(false);
+  const [isMintingObservatory, setIsMintingObservatory] = useState(false);
+  const [observatoryMintError, setObservatoryMintError] =
```
Comment (adamhannigan) [inline src/features/goblins/rocket/Rocket.tsx:49]: When a component has 4 different states it is usually a good indicator to split the functionality or use a state management solution like xstate.

To help maintain this complex state I would suggest introducing a separate business logic handler or state machine.

File: src/features/goblins/rocket/Rocket.tsx [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+    setIsRocketLaunching(true);
+  };
+
+  const handleOpenDialog = () => {
+    setIsDialogOpen(true);
+    if (
+      isMelonDuskSeen &&
+      !isRocketFixed &&
+      !hasCompletedMission &&
+      !hasCompletedQuest
```
Comment (adamhannigan) [inline src/features/goblins/rocket/Rocket.tsx:97]: This is a good example where having too many states in a component becomes hard to manage.

If a new state was to be added, this component would get even more complex and increase the chance of a regression

File: src/features/goblins/rocket/Rocket.tsx [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+  };
+
+  const handleCloseDialog = () => {
+    melonDuskAudio.stop();
+    setIsDialogOpen(false);
+    setObservatoryMintError(null);
+  };
+
+  const handleOpenItemsDialog = () => {
+    localStorage.setItem(MELON_DUSK_SEEN, JSON.stringify(true));
```
Comment (adamhannigan) [inline src/features/goblins/rocket/Rocket.tsx:113]: I would suggest abstracting local storage functionality out of the component. Where possible, a component should only reflect UI behaviour. Business logic + data storage can be more maintainable in separate files

File: src/features/goblins/rocket/Rocket.tsx [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    setIsItemsOpen(true);
+  };
+
+  const handleMintObservatory = async () => {
+    setIsMintingObservatory(true);
+    try {
+      await metamask.getMillionOnMars().trade();
```
Comment (adamhannigan) [inline src/features/goblins/rocket/Rocket.tsx:120]: Consider calling business functionality and actions outside of the UI layer (components)

**Review by adamhannigan — APPROVED** [review]
Awesome work team ❤️ Can't wait for the players to see this!

One suggestion is around the architecture of the Rocket.tsx component. The Modals, sound FX and animations are on point 👌 

Where possible, the responsibility of a React component should be purely UI and reflecting the current application state. The Rocket component is trying to maintain too many states, is performing business logic and also handling data storage. This makes the component difficult to reason about, harder to maintain and more chance of a fatal bug appearing.

The change I am proposing is following the approach other areas in the game use for heavy state management: xstate 🚀 I understand this is quite a big change given our timeline for release and is out of scope for this PR.

I will merge this in and open a PR to help refactor the logic/workflows out of the UI layer

**Comment by ShiningMinions** [comment]
@adamhannigan great suggestions, and I 100% agree with all your points. I think it's worth noting that we were focused on speed of release over cleaner code (because of your tight deadline). Now that we have extra time we can work on cleaning up these extra bits.

My only rebuttal: isn't this supposed to be a temporary event? I think one valuable thing about it all being in one component, is the ease of removal when the event ends.

**Comment by adamhannigan** [comment]
@ShiningMinions It is temporary, but we want to be consistent in development ;) 

Once we start taking shortcuts it is hard to go back

**Comment by adamhannigan** [comment]
Screen Shot 2022-05-23 at 10 02 35 am

@ShiningMinions Can you please share the source of this music. We cannot merge anything in until the licence is confirmed. If we use the wrong licence it could be the end of our entire project.

**Comment by ShiningMinions** [comment]
> Screen Shot 2022-05-23 at 10 02 35 am
> 
> @ShiningMinions Can you please share the source of this music. We cannot merge anything in until the licence is confirmed. If we use the wrong licence it could be the end of our entire project.

Yup, so I didn't create these, I just used them. I can tell you who I got them from:

- rocket_launch.mp3 was from AetherZapdos - I believe it was clipped from youtube so you should probably get this one sorted out.
- SFL_in_Space.mp3 was provided by RadishPies - it was assumed that this was an original tune they made, but I am not certain.

You should talk to @Z1kou about this, probably best person to talk to.

**Pull request merged**